openedx/core/djangoapps/programs/utils.py

```diff
@@ -1,49 +1,105 @@
-"""
-Helper methods for Programs.
-"""
+"""Helper functions for working with Programs."""
+import logging
+
 from django.core.cache import cache
 from edx_rest_api_client.client import EdxRestApiClient
+
 from openedx.core.djangoapps.programs.models import ProgramsApiConfig
+from openedx.core.lib.token_utils import get_id_token
 
 
-def is_student_dashboard_programs_enabled():    # pylint: disable=invalid-name
-    """ Returns a Boolean indicating whether LMS dashboard functionality
-     related to Programs should be enabled or not.
-    """
-    return ProgramsApiConfig.current().is_student_dashboard_enabled
+log = logging.getLogger(__name__)
 
 
-def programs_api_client(api_url, jwt_access_token):
-    """ Returns an Programs API client setup with authentication for the
-    specified user.
-    """
-    return EdxRestApiClient(
-        api_url,
-        jwt=jwt_access_token
-    )
+def get_programs(user):
+    """Given a user, get programs from the Programs service.
 
+    Returned value is cached depending on user permissions. Staff users making requests
+    against Programs will receive unpublished programs, while regular users will only receive
+    published programs.
 
-def is_cache_enabled_for_programs():    # pylint: disable=invalid-name
-    """Returns a Boolean indicating whether responses from the Programs API
-    will be cached.
+    Arguments:
+        user (User): The user to authenticate as when requesting programs.
+
+    Returns:
+        list of dict, representing programs returned by the Programs service.
     """
-    return ProgramsApiConfig.current().is_cache_enabled
+    programs_config = ProgramsApiConfig.current()
+    cache_key = programs_config.get_cache_key(user)
+    no_programs = []
+
+    if not programs_config.enabled:
+        log.warning('Programs configuration is disabled.')
+        return no_programs
+
+    if programs_config.is_cache_enabled:
+        cached = cache.get(cache_key)
+        if cached:
```

Comment:
i think this needs to differentiate whether the cached value was an empty list (vs being `None`) - otherwise the cache will be bypassed while the result was empty